In Anaconda prompt, enter 'pip install lxml' and 'pip install requests'

In [10]:
from lxml import html
import requests
import pandas as pd
import numpy as np
import matplotlib
import matplotlib.pyplot as plt
matplotlib.style.use('ggplot')
%matplotlib inline
import math
from itertools import groupby
import collections

## Insight Health Data Science Fellows

### This program currently operates on site in either Boston, MA or Seattle, WA (no Remote option)

In [30]:
page = requests.get('http://insighthealthdata.com/fellows')
tree = html.fromstring(page.content)

Which companies have hired the most Health fellows?

In [31]:
#Extract a list of names:
health_name = tree.xpath('//div[@class="fellow_name"]/text()')
#Extract a list of companies:
health_company = tree.xpath('//div[@class="fellow_company"]/text()')
#Extract job title:
health_title = tree.xpath('//div[@class="toottip_title"]/text()')
#Extract academic background: 
health_bckgrd = tree.xpath('//div[@class="tooltip_background"]/text()')
health_project = tree.xpath('//div[@class="tooltip_project"]/text()')

In [32]:
counter=collections.Counter(health_company)
print(counter.most_common(6))
    

[('Seven Bridges', 9), ('Wayfair', 8), ('Tamr', 3), ('Insight Data Science', 3), ('Amazon', 3), ('The Broad Institute', 2)]


**I found it surprising that Wayfair was such a common hirer of Health Data Science Fellows. Upon further research, they have an office in Boston so it makes sense.** 

In [33]:
health_jobs=pd.DataFrame(data=[health_title]).transpose()
health_jobs.columns=['Job title']
title_freq = health_jobs.apply(pd.value_counts)
title_freq[:20]

Job title
Data Scientist                                  29
Data Scientist                                  11
R&D Scientist                                    3
Computational Biologist                          3
Field Engineer                                   3
Scientific Project Manager                       3
Program Director, Data Scientist                 2
Bioinformatician                                 2
Principal Scientist I                            2
Data Scientist, Speech                           2
Consultant                                       2
Scientist                                        2
Senior Statistician                              1
Senior Data Engineer Associate                   1
Program Director, Data Science                   1
Program Director, Data Science                   1
Analyst, Health Analytic Solutions               1
Bioinformatics Scientist                         1
Quantitative User Experience Researcher          1
Business Intelligence Consultant                 1

In [34]:
health_bck=pd.DataFrame(data=[health_bckgrd]).transpose()
health_bck.columns=['background']
health_bck[:20]

background
0         Neuroscience, Johns Hopkins University, PhD
1                   Systems Biology, Harvard, Postdoc
2      Cognitive Neuroscience, McGill University, PhD
3                      Cell Biology, Harvard, Postdoc
4   Bioinformatics, University of Massachusetts Me...
5            Environmental Sciences, Harvard, Postdoc
6     Infectious Disease, Cornell University, Postdoc
7   Immunology & Infectious Disease, Harvard, Postdoc
8           Astrophysics, Cornell University, Postdoc
9                           Biology, Harvard, Postdoc
10  Computational Biology, University of Texas Sou...
11              Bioinformatics/Genomics, MIT, Postdoc
12               Psychology, Duke University, Postdoc
13  Sleep & Cardiovascular Medicine, Brigham & Wom...
14                   Physics, Cornell University, PhD
15                              Physics, Harvard, PhD
16       Bioengineering, University of San Diego, PhD
17        Biophysics, University of Michigan, Postdoc
18  Neuroscience, University of Pennsylvania, Postdoc
19           Neuroscience, Temple University, Postdoc

## <font color=blue> Insight Data Science Fellows </font>

### Scrape Fellow data from webpage

In [11]:
page_2 = requests.get('http://insightdatascience.com/fellows')
tree_2 = html.fromstring(page_2.content)
#Extract Fellow name:
fellow_name=tree_2.xpath('//div[@class="fellow_name"]/text()')
#Extract a list of companies:
fellow_company = tree_2.xpath('//div[@class="fellow_company"]/text()')
#Extract job title:
fellow_title = tree_2.xpath('//div[@class="toottip_title"]/text()')
#Extract academic background: 
fellow_bckgrd = tree_2.xpath('//div[@class="tooltip_background"]/text()')
#Extract project title/description
fellow_project = tree_2.xpath('//div[@class="tooltip_project"]/text()')

In [13]:
page_2 = requests.get('http://insightdatascience.com/fellows')
tree_2 = html.fromstring(page_2.content)
name = tree_2.xpath('//div[@class="fellow_name"]/text()')
company = tree_2.xpath('//div[@class="fellow_company"]/text()')
title = tree_2.xpath('//div[@class="toottip_title"]/text()')
background = tree_2.xpath('//div[@class="tooltip_background"]/text()')
project = tree_2.xpath('//div[@class="tooltip_project"]/text()')

In [29]:
len_name = len(fellow_name)
print("There are %a Insight Data Science Fellows featured on the fellows page." %len_name )

There are 702 Insight Data Science Fellows featured on the fellows page.


Which companies have hired the most Data Science Fellows?

In [14]:
counter_2 = collections.Counter(company)
print(counter_2.most_common(5))

[('Facebook', 46), ('Stitch Fix', 19), ('LinkedIn', 16), ('Netflix', 13), ('Insight Data Science', 13)]


Top five companies that have hired Insight Data Science Fellows

Facebook: 46

Stitch Fix: 19

LinkedIn: 16

Netflix: 13

Insight Data Science: 11

How many unique companies have hired Insight Data Science Fellows?

In [15]:
len(company)
companyset = set(company)
len(companyset)

345

There are 345 unique companies that have hired Insight Data Science Fellows; 702 fellows were included in this dataset

Examine all companies that have hired fellows

In [26]:
df=pd.DataFrame(data=[company]).transpose()
df.columns=['company']
company_freq = df.apply(pd.value_counts)
company_freq[:30]


company
Facebook                          46
Stitch Fix                        19
LinkedIn                          16
Insight Data Science              13
Netflix                           13
Gartner                           12
Intuit                            12
Yelp                              11
Square                             9
Schireson                          9
Capital One                        9
Airbnb                             9
Uber                               8
Twitter                            8
Silicon Valley Data Science        7
Google                             7
Viacom                             6
Microsoft Yammer                   6
Apple                              6
Grand Rounds                       5
Bloomberg                          5
IBM                                5
SalesforceIQ                       5
Nielsen                            5
Macy's                             5
Amazon                             4
Twitch                             4
Sojern                             4
Vanguard                           4
Tamr                               4

Which fellows hired by Insight Data Science?

In [17]:
df_2 = pd.DataFrame(data=[company, title]).transpose()
df_2.columns=['Company','Title']
df_2[:10]

Company                           Title
0     Facebook                  Data Scientist
1        Adobe                  Data Scientist
2     LinkedIn  Head of Anti-Abuse Engineering
3  Capital One                Director of Risk
4   Proofpoint                  Data Scientist
5     LinkedIn           Senior Data Scientist
6      Twitter                  Data Scientist
7      Netflix                  Data Scientist
8     Facebook                  Data Scientist
9     LinkedIn           Senior Data Scientist

In [20]:
df_2[df_2.Company == 'Insight Data Science']

Company                              Title
75   Insight Data Science                Director of Product
130  Insight Data Science                Director of Product
228  Insight Data Science                Director of Product
329  Insight Data Science       Artificial Intelligence Lead
371  Insight Data Science  Program Director & Data Scientist
400  Insight Data Science  Program Director & Data Scientist
439  Insight Data Science       Artificial Intelligence Lead
476  Insight Data Science  Program Director & Data Scientist
484  Insight Data Science  Program Director & Data Scientist
485  Insight Data Science  Program Director & Data Scientist
522  Insight Data Science  Program Director & Data Scientist
536  Insight Data Science  Program Director & Data Scientist
586  Insight Data Science                   Program Director

In [23]:
df_3=pd.DataFrame(data=[background]).transpose()
df_3.columns=['background']
background_freq = df_3.apply(pd.value_counts)
background_freq[:10]

background
Neuroscience, New York University, Postdoc                   5
Biomedical Engineering, Johns Hopkins Universit...           4
Applied Physics, Stanford, PhD                               4
Physics, Stony Brook University, PhD                         4
Policy Analysis, Pardee RAND Graduate School, PhD            3
Neuroscience, UC Berkeley, Postdoc                           3
Materials Science & Engineering, Cornell Univer...           2
Physics, University of Toronto, PhD                          2
Applied Mathematics, Brown University, PhD                   2
Computer Science, University of British Columbi...           2

Under what titles are Fellows hired?

In [22]:
df_2=pd.DataFrame(data=[title]).transpose()
df_2.columns=['title']
title_freq = df_2.apply(pd.value_counts)
title_freq[:20]

title
Data Scientist                       367
Senior Data Scientist                 42
Data Scientist                        13
Software Engineer                     11
Lead Data Scientist                    8
Data Engineer                          8
Sr. Data Scientist                     8
Program Director & Data Scientist      7
Business Associate                     7
Data Science Consultant                6
Research Scientist                     5
Consultant, Data Science               4
Machine Learning Engineer              4
Quantitative Researcher                4
Quantitative Analyst                   4
Field Engineer                         3
Director of Product                    3
Analytics Manager                      3
Operations Analyst                     3
Security Data Scientist                3

There may be some spacing issues, since "Data Scientist" appears twice

In [25]:
background

['Bio-X Games, Stanford, Postdoc',
 'Space Physics, Virgina Tech, PhD',
 'Physics, Harvard University,\xa0PhD\n',
 'Operations Research, MIT, PhD',
 'Physics, UC Riverside, PhD',
 'Mathematics, Stanford, Postdoc',
 'Mathematics, Duke, Postdoc',
 'Statistics, Stanford, Postdoc',
 'Applied Mathematics, UC Berkeley, PhD',
 'Computational & Mathematical Engineering, Stanford, PhD',
 'Space Plasma Physics, Stanford, Postdoc',
 'Electrical Engineering & Computer Science, UC Berkeley, PhD',
 'Computational & Mathematical Engineering, Stanford, PhD',
 'Computational Biology, UC Berkeley, Postdoc',
 'Electrical Engineering & Computer Science, UC Berkeley, Postdoc',
 'Computer Vision & Image Processing, Johns Hopkins University, Postdoc',
 'Environmental Energy, UC Berkeley, Postdoc',
 'Computational Biophysics, Harvard University, Postdoc',
 'Biology, Stanford, PhD',
 'Computational Neuroscience, UC Berkeley, Postdoc',
 'Condensed Matter Physics,\xa0Lawrence Berkeley National Laboratory, Postdo

Each Fellow has a string of data containing their major, school, and degree status. Need to break this up into three separate variables for each Fellow.

Iterate over each string in the list, then split each string into three separate elements while preserving that the original string belonged to one Fellow. 

In [29]:
bck = []
for x in background:
    bck.append(x.split(","))


In [30]:
df_4 = pd.DataFrame(bck)
df_4.columns = ['Major', 'School', 'Status',"Col4","Col5"]
df_4

Major  \
0                                  Bio-X Games   
1                                Space Physics   
2                                      Physics   
3                          Operations Research   
4                                      Physics   
5                                  Mathematics   
6                                  Mathematics   
7                                   Statistics   
8                          Applied Mathematics   
9     Computational & Mathematical Engineering   
10                        Space Plasma Physics   
11   Electrical Engineering & Computer Science   
12    Computational & Mathematical Engineering   
13                       Computational Biology   
14   Electrical Engineering & Computer Science   
15          Computer Vision & Image Processing   
16                        Environmental Energy   
17                    Computational Biophysics   
18                                     Biology   
19                  Computational Neuroscience   
20                    Condensed Matter Physics   
21                    Mathematical Engineering   
22                           Molecular Imaging   
23                                    Genetics   
24                                Neuroscience   
25               Experimental Particle Physics   
26                            Particle Physics   
27                                     Physics   
28               Experimental Particle Physics   
29                        Space Plasma Physics   
..                                         ...   
673              Experimental Particle Physics   
674                           Education Policy   
675                             Plasma Physics   
676                               Neuroscience   
677                                  Materials   
678                                    Physics   
679                              Solar Physics   
680              Experimental Particle Physics   
681                       Cognitive Psychology   
682                               Neuroscience   
683                                  Astronomy   
684        Mathematical & Statistical Sciences   
685                                   Politics   
686     Electrical and Electronics Engineering   
687            Materials Science & Engineering   
688                         Structural Biology   
689                          Political Science   
690                       Nuclear Astrophysics   
691                                    Physics   
692                       Department of Energy   
693         Cognitive and Information Sciences   
694              Experimental Particle Physics   
695                     Cognitive Neuroscience   
696                     Mechanical Engineering   
697                       Systems Neuroscience   
698                          Molecular Biology   
699                            Systems Biology   
700                     Electrical Engineering   
701            Theoretical High Energy Physics   
702                                    Physics   

                                     School    Status  Col4  Col5  
0                                  Stanford   Postdoc  None  None  
1                              Virgina Tech       PhD  None  None  
2                        Harvard University     PhD\n  None  None  
3                                       MIT       PhD  None  None  
4                              UC Riverside       PhD  None  None  
5                                  Stanford   Postdoc  None  None  
6                                      Duke   Postdoc  None  None  
7                                  Stanford   Postdoc  None  None  
8                               UC Berkeley       PhD  None  None  
9                                  Stanford       PhD  None  None  
10                                 Stanford   Postdoc  None  None  
11                              UC Berkeley       PhD  None  None  
12                                 Stanford       PhD  N

In [90]:
pd.value_counts(df_4.Major)

Physics                                        69
Astrophysics                                   43
Neuroscience                                   25
Mathematics                                    23
Particle Physics                               16
Cognitive Neuroscience                         14
Computational Biology                          12
Biomedical Engineering                         12
Biophysics                                     11
Computer Science                               10
Electrical Engineering                          9
Applied Physics                                 9
Astronomy                                       8
Applied Mathematics                             8
Bioinformatics                                  8
Mechanical Engineering                          8
Computational Neuroscience                      7
Experimental Particle Physics                   7
Systems Biology                                 7
Statistics                                      6


Instead of producing just three columns, there are five columns. So there must be multiple commas for some fellows' information. Ex: University of California, Los Angeles should remain one string but would be split using my command.


In [31]:
pd.value_counts(df_4.Status)

 PhD                                           402
 Postdoc                                       183
 PhD                                            16
 Postdoc                                         5
 Faculty                                         4
 Associate Research Scientist                    3
 Boulder                                         3
 Assistant Professor                             3
 Berkeley                                        2
 Riverside                                       2
 Lecturer                                        2
 Santa Barbara                                   2
 College Park                                    2
 Staff                                           2
 San Diego                                       2
 Austin                                          2
 Los Angeles                                     2
 Australia                                       2
 Research Associate                              2
 PhD                           

In [32]:
def phd(series):
    if series == 'PhD\n':
        return 'PhD'
    elif series == 'PhD\n ':
        return 'PhD'
    elif series == 'PhD ':
        return 'PhD'    
    else:
        return series
df_4['Status'] = df_4['Status'].apply(phd)
        
df_4['Status'].value_counts()

 PhD                                           402
 Postdoc                                       183
 PhD                                            16
 Postdoc                                         5
 Faculty                                         4
 Associate Research Scientist                    3
 Boulder                                         3
 Assistant Professor                             3
 Berkeley                                        2
 Riverside                                       2
 Lecturer                                        2
 Santa Barbara                                   2
 College Park                                    2
 Staff                                           2
 San Diego                                       2
 Austin                                          2
 Los Angeles                                     2
 Australia                                       2
 Research Associate                              2
 PhD                           

In [95]:
pd.value_counts(df_4.School)

 Stanford                                                                49
 UC Berkeley                                                             44
 Stanford University                                                     21
 New York University                                                     17
 University of California                                                12
 Columbia University                                                     11
 Harvard                                                                 11
 UC Davis                                                                 9
 Johns Hopkins University                                                 9
 University of Pennsylvania                                               9
 MIT                                                                      9
 University of Illinois at Urbana-Champaign                               8
 UCLA                                                                     7
 Rutgers Uni

Interesting finding... There are additional commas which cause the fellow's major or name of a university to extend into the next column. In general, most Fellows enter the program with a PhD or as a Postdoc

In [36]:
project

['Measuring the Impact of Open Source Coders on Github',
 'CouchTube: YouTube TV shows in one click',
 'Email Searcher: Search & Visualize Your Mailbox',
 'Tweet Timeline: Visualizing the Impact of Social Media',
 'Analysis of New York Stop-and-Frisk Data',
 'Flight Delay Predictor',
 'Email Contacts Automatically Grouped & Ranked',
 'Sentiment Analysis Based on Social Media',
 'Sort My Friends: Automatically Organize your Connections',
 'Diverse Recipe Recommendations',
 'Arrive in Time: Airport Arrival Time Predictor',
 'Flat Finder: Find Your Perfect Apartment',
 'Stock Volatility Predictor\xa0',
 'SchoolGeo: Empowering Parents to Make Informed Decisions',
 'Event Map: Get Local News from Twitter',
 'Multi-Domain Recommendations',
 'Lower Your Electricity Bill',
 'Find the Expert on Stack Overflow',
 'STARtrack: Helping Educators Track Student Performance',
 'Kiva Loan Checker',
 'FlockThink: Twitter Search Engine For Instant Sentiment Analysis\xa0',
 'ReadTweetApp: The Tweets That 

In [35]:
 df_4.to_excel('C:/Users/sugac_000/Desktop/Python SQL/Insight Data Project/InsightDataSci.xlsx', sheet_name='Data')

I exported the existing, messy dataset to an Excel file to do some additional cleaning.

In [3]:
df_5 = pd.read_excel('C:/Users/sugac_000/Desktop/Python SQL/Insight Data Project/InsightDataSci.xlsx', sheet_name='Data')

In [4]:
df_5[:10]

Major          Institution   Status  \
0                               Bio-X Games             Stanford  Postdoc   
1                             Space Physics         Virgina Tech      PhD   
2                                   Physics   Harvard University      PhD   
3                       Operations Research                  MIT      PhD   
4                                   Physics         UC Riverside      PhD   
5                               Mathematics             Stanford  Postdoc   
6                               Mathematics                 Duke  Postdoc   
7                                Statistics             Stanford  Postdoc   
8                       Applied Mathematics          UC Berkeley      PhD   
9  Computational & Mathematical Engineering             Stanford      PhD   

  Country  
0     USA  
1     USA  
2     USA  
3     USA  
4     USA  
5     USA  
6     USA  
7     USA  
8     USA  
9     USA

In [5]:
pd.value_counts(df_5.Institution)

 Stanford                                            49
 UC Berkeley                                         47
 Stanford University                                 28
 New York University                                 22
 Columbia University                                 14
 Harvard                                             11
 UC Davis                                            11
 Rutgers University                                  10
 MIT                                                 10
 University of Pennsylvania                          10
 Johns Hopkins University                            10
 Harvard University                                   9
 Stony Brook University                               9
 University of Illinois at Urbana-Champaign           9
 University of Toronto                                8
 Lawrence Berkeley National Laboratory                7
 California Institute of Technology                   7
 UCLA                                           

In [21]:
pd.value_counts(df_5.Country)

USA             610
Canada           29
UK               16
Australia         7
Switzerland       7
France            4
Germany           3
Netherlands       2
Spain             2
South Africa      1
Taiwan            1
Kenya             1
Denmark           1
Singapore         1
Israel / USA      1
Japan             1
Turkey            1
Belgium           1
Italy             1
New Zealand       1
Name: Country, dtype: int64

In [23]:
610/703*100

86.7709815078236